### PR TITLE
[rllib] Fix logging to Athena

### PR DESCRIPTION
**Pull request by pcmoritz**
This fixes two problems the came up when writing logs to Athena:

1. TensorFlow doesn't support logging to S3 (we are falling back to /tmp now)
2. There were NaNs that were floats and not np.float so our JSONEncoder didn't catch them (it is only invoked for unknown objects)

### CODE REVIEW

File: python/ray/rllib/a3c/a3c.py [hunk 1]
```diff
@@ -73,8 +73,13 @@ def get_completed_rollout_metrics(self):
         return completed
 
     def start(self):
+        logdir = self.logdir
+        if logdir.startswith("s3"):
+            print("WARNING: TensorFlow logging to S3 not supported by TensorFlow,"
```
Comment (richardliaw) [inline python/ray/rllib/a3c/a3c.py]: (lint)

File: python/ray/rllib/ppo/ppo.py [hunk 2]
```diff
@@ -99,8 +99,13 @@ def _init(self):
             for _ in range(self.config["num_workers"])]
         self.start_time = time.time()
         if self.config["write_logs"]:
+            logdir = self.logdir
+            if logdir.startswith("s3"):
+                print("WARNING: TensorFlow logging to S3 not supported by TensorFlow,"
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: small lint

File: python/ray/rllib/a3c/a3c.py [hunk 3]
```diff
@@ -73,8 +73,13 @@ def get_completed_rollout_metrics(self):
...
+                  " logging to /tmp/ instead")
+            logdir = "/tmp/"
         summary_writer = tf.summary.FileWriter(
-            os.path.join(self.logdir, "agent_%d" % self.id))
+            os.path.join(logdir, "agent_%d" % self.id))
```
Comment (richardliaw) [inline python/ray/rllib/a3c/a3c.py]: [1] In `common.py`, the tmp directory is a `/tmp/ray/ + prefix`. It makes sense to make that consistent and also to abstract all the logging file naming code into a separate file, perhaps in a separate PR.

File: python/ray/rllib/ppo/ppo.py [hunk 4]
```diff
@@ -99,8 +99,13 @@ def _init(self):
...
+                      " logging to /tmp/ instead")
+                logdir = "/tmp/"
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: Same as [1]

File: python/ray/rllib/common.py [hunk 5]
```diff
@@ -21,13 +21,37 @@
...
+        self.nan_str = nan_str
+
+    def iterencode(self, o, _one_shot=False):
+        if self.ensure_ascii:
+            _encoder = json.encoder.encode_basestring_ascii
+        else:
+            _encoder = json.encoder.encode_basestring
+
+        def floatstr(o, allow_nan=self.allow_nan, nan_str=self.nan_str):
+            if o != o:
```
Comment (richardliaw) [inline python/ray/rllib/common.py]: `o` is certainly equal to `o`; perhaps name mismatch? Also, `allow_nan` is not used.

Also, this logic can probably be readably condensed into something like 1 or 2 lines.

**Review by richardliaw — CHANGES_REQUESTED** [review]
A few changes to be made

File: python/ray/rllib/common.py (hunk 5 above)
Comment (pcmoritz) [reply python/ray/rllib/common.py]: o != o if o is NaN, so it is correct actually ;)

will try to condense

**Review by richardliaw — APPROVED** [review]

**Pull request merged**